examples/powertools-examples-cloudformation/src/main/java/helloworld/App.java

```diff
@@ -0,0 +1,166 @@
+        int retentionDays = Integer.parseInt(String.valueOf(cloudFormationCustomResourceEvent.getResourceProperties().get("RetentionDays")));
+        log.info("Retention Period (Days) {}", retentionDays);
+        String newBucketName = (String) cloudFormationCustomResourceEvent.getResourceProperties().get("BucketName");
+
+        if (!physicalResourceId.equals(newBucketName)) {
+            // Custom Resource is configured with a different bucket name
+            try {
+                createBucketWithRetention(newBucketName, retentionDays);
+                return Response.success(newBucketName);
+            } catch (AwsServiceException | SdkClientException e) {
+                log.error(e);
+                return Response.failed(newBucketName);
+            }
+        } else {
+            // Bucket name has not changed - check for changes in Lifecycle configuration
+            try {
+                updateLifecycleRules(physicalResourceId, retentionDays);
+                return Response.success(physicalResourceId);
+            } catch (AwsServiceException | SdkClientException e) {
+                log.error(e);
+                return Response.failed(physicalResourceId);
+            }
+        }
+    }
+
+    @Override
+    protected Response delete(CloudFormationCustomResourceEvent cloudFormationCustomResourceEvent, Context context) {
+        Objects.requireNonNull(cloudFormationCustomResourceEvent, "cloudFormationCustomResourceEvent cannot be null.");
+        Objects.requireNonNull(cloudFormationCustomResourceEvent.getPhysicalResourceId(), "PhysicalResourceId cannot be null.");
+
+        log.info(cloudFormationCustomResourceEvent);
+        String bucketName = cloudFormationCustomResourceEvent.getPhysicalResourceId();
+        log.info("Bucket Name {}", bucketName);
+
+        if (bucketExists(bucketName)) {
+            try {
+                s3Client.deleteBucket(DeleteBucketRequest.builder().bucket(bucketName).build());
+                log.info("Bucket Deleted {}", bucketName);
+                return Response.success(bucketName);
+            } catch (AwsServiceException | SdkClientException e) {
+                log.error(e);
+                return Response.failed(bucketName);
+            }
+        } else {
+            log.info("Bucket already deleted - no action");
+            return Response.success(bucketName);
+        }
+
+    }
+
+    private boolean bucketExists(String bucketName) {
+        try {
+            HeadBucketResponse headBucketResponse = s3Client.headBucket(HeadBucketRequest.builder().bucket(bucketName).build());
+            if (headBucketResponse.sdkHttpResponse().isSuccessful()) {
+                return true;
+            }
+        } catch (NoSuchBucketException e) {
+            log.info("Bucket does not exist");
+            return false;
+        }
+        return false;
+    }
+
+    private void updateLifecycleRules(String physicalResourceId, int retentionDays) {
```

Comment:
Yes, it looks a bit complicated for a sample. Maybe simply use the one from the doc (https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/walkthrough-custom-resources-lambda-lookup-amiids.html), so you can even reference it.

Reply:
I think the example in the docs is too trivial because it does not create any underlying AWS resources.
I wanted to have a more complete example which showed how to handled all the scenarios.

Happy to refactor the code, adding comments and some more info on the `README`. Would it be enough?

Reply:
@mriccia @jeromevdl I tend to agree that the example there is a poor fit - it's using a custom resource to fetch some data, rather than creating a resource itself, which i'd say is 1/ less common, or more of a special case and 2/ doesn't highlight the "gotchas" of the custom resource model.

I still think the App.java impl needs to be quite explicitly documented (see other comment - "there's a bit to mentally unpack"); IMHO the example should make it super clear _why_ you need to do this, like delete/recreate if properties change - so that people that work from it understand what's happening and don't get bitten by weird lifecycle effects later on.